app/code/Magento/Catalog/Block/Product/View.php

```diff
@@ -173,7 +181,12 @@ public function getProduct()
             $product = $this->productRepository->getById($this->getProductId());
             $this->_coreRegistry->register('product', $product);
         }
-        return $this->_coreRegistry->registry('product');
+        $product = $this->_coreRegistry->registry('product');
+        if (! $product) {
```

Comment:
@orlangur I'll check if its unnescessary and remove the space its just a personal preference for readability